### PR TITLE
Add `broadcast!(identity, ::Array, ::Array)` specialization

### LINKED ISSUE #20227: Add broadcast!{S,T,N}(identity, A::Array{S,N}, B::Array{T,N}) specialization calling memcpy
#19895 removed an (incorrect and ambiguity-causing) `broadcast!` specialization (roughly `broadcast!{S,T,N}(identity, A::AbstractArray{S,N}, B::AbstractArray{T,N}) = copy!(A, B)`). Restoring a narrower, corrected form of that specialization for `Array`s might be worthwhile; see discussion towards the end of #19895. `@martinholters` sketched an implementation [here](https://github.com/JuliaLang/julia/pull/19895#issuecomment-272386502). (@martinholters, do you plan to wrap that into a PR, or should we mark this up for grabs?) Best!

### PR DESCRIPTION
**Pull request by martinholters**
Closes #20227.

Performance improvement is most prominent for small element size:
```julia
x=rand(Int8,50,50); y=similar(x); @benchmark $y .= $x

# master
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     9
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     2.20 μs (0.00% GC)
  median time:      2.21 μs (0.00% GC)
  mean time:        2.22 μs (0.00% GC)
  maximum time:     3.38 μs (0.00% GC)

# this PR
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     976
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     70.00 ns (0.00% GC)
  median time:      70.00 ns (0.00% GC)
  mean time:        70.19 ns (0.00% GC)
  maximum time:     130.00 ns (0.00% GC)
```

There is a little bit of overhead though, becoming significant for very small arrays:
```julia
x=rand(Int64,1); y=similar(x); @benchmark $y .= $x

# master
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     999
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     9.00 ns (0.00% GC)
  median time:      10.00 ns (0.00% GC)
  mean time:        9.96 ns (0.00% GC)
  maximum time:     19.00 ns (0.00% GC)

# this PR
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     998
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     15.00 ns (0.00% GC)
  median time:      15.00 ns (0.00% GC)
  mean time:        15.21 ns (0.00% GC)
  maximum time:     34.00 ns (0.00% GC)
```

### CODE REVIEW

File: base/broadcast.jl [hunk 1]
```diff
@@ -15,6 +15,13 @@ typealias ScalarType Union{Type{Any}, Type{Nullable}}
 # fallbacks for some special cases
 @inline broadcast(f, x::Number...) = f(x...)
 @inline broadcast{N}(f, t::NTuple{N,Any}, ts::Vararg{NTuple{N,Any}}) = map(f, t, ts...)
+function broadcast!{T,S,N}(::typeof(identity), x::Array{T,N}, y::Array{S,N})
+    if size(x) == size(y)
+        copy!(x, y)
+    else
+        Broadcast.broadcast_c!(identity, Array, Array, x, y)
```
Comment (Sacha0) [inline base/broadcast.jl]: This definition living within the `Broadcast` module, no need for the qualification?

File: base/broadcast.jl [hunk 2]
```diff
@@ -15,6 +15,13 @@ typealias ScalarType Union{Type{Any}, Type{Nullable}}
...
+    end
+end
```
Comment (Sacha0) [inline base/broadcast.jl]: Perhaps reduce this definition from seven to two lines?
```julia
broadcast!{T,S,N}(::typeof(identity), x::Array{T,N}, y::Array{S,N}) =
    size(x) == size(y) ? copy!(x, y) : broadcast_c!(identity, Array, Array, x, y)
```
Best!

**Review by Sacha0 — APPROVED** [review]
LGTM! Please see two minor review comments. Best!

**Comment by martinholters** [comment]
Replaced the new specialization with the shortened version proposed by @Sacha0.

**Pull request merged**